### PR TITLE
DEV: Adds gitpod to numpy

### PR DESCRIPTION
**Pull request by thomasjpfan**
Related to https://github.com/scipy/scipy/issues/13677. This PR adds gitpod integration to numpy. Here is a [gitpod link to my fork](https://gitpod.io/#https://github.com/thomasjpfan/numpy/tree/gitpod) so you can try it out. You should see that numpy builds quickly because it is using `ccache`.

- Uses a prebuilt image that builds numpy with ccache enabled. Forks can use the same `.gitpod.yml` and quickly build numpy when their workspace starts. 
- Uses an image I built and pushed to [docker hub](https://hub.docker.com/repository/docker/thomasjpfan/numpy-gitpod). Moving forward, the image should be built with a github action and updated to docker hub.
- A bare minimum configuration to get everything setup at first. (no extensions/doc building etc.)
- Uses `gitpod/workspace-base` as a base layer because `gitpod` will have it downloaded in their system already.

#### Update
- Uses `gitpod/workspace-base` now since conda is used to install Python + the complete environment
- Setting up ccache is a little more involved with conda-forge compilers

CC @rgommers

### CODE REVIEW

File: tools/gitpod/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,20 @@
+FROM gitpod/workspace-full
+
+RUN sudo apt-get update && \
+    sudo apt-get install --no-install-recommends -y ccache gfortran && \
+    sudo rm -rf /var/lib/apt/lists/*
```
Comment (rgommers) [inline tools/gitpod/Dockerfile:5]: Could we use conda like SciPy does? It's much easier to get a complete dev environment that way. Can use the `environment.yml` in the root of the repo for it.

**Comment by rgommers** [comment]
Thanks for working on this @thomasjpfan!

Cc @trallard, @melissawm who were also planning on working on Gitpod for NumPy.

**Comment by rgommers** [comment]
> Uses an image I built and pushed to [docker hub](https://hub.docker.com/repository/docker/thomasjpfan/numpy_gitpod). Moving forward, the image should be built with a github action and updated to docker hub.

Guess we'll want a `numpy` org on Docker Hub. Or we can be a bit lazy and use https://hub.docker.com/repository/docker/scipy/scipy-dev. Not sure about whether it matters to have more images in a single org or not.

File: tools/gitpod/Dockerfile (hunk 1 above)
Comment (thomasjpfan) [reply tools/gitpod/Dockerfile]: I updated the PR to use `environment.yml`. The extra dependencies does increase the size of the  docker image, which makes the "Pulling docker image" screen take around ~1 minute.

**Comment by thomasjpfan** [comment]
Update PR with gitpod prebuilts, a github action to only build the Dockerfile, using `environment.yml` with conda and using a smaller base image: `gitpod/workspace-base`.

File: tools/gitpod/Dockerfile (hunk 1 above)
Comment (rgommers) [reply tools/gitpod/Dockerfile]: I think that's still worth it, missing dependencies that are necessary to work on docs or type annotations for example would be worse, because those are tasks that are good for new contributors. And a dev env without IPython is seriously incomplete imho:)

**Comment by trallard** [comment]
I am finishing the SciPy gitpod rework today, I had a quick look at the changes here and decided against starting a review to avoid lengthy comments

Some things that need rethinking or having in mind:
- It is worth using conda like Ralf mentioned -> eventually using the Docker cache in GH actions should help with the image creation times and definitely makes the build and debugging easier
- Permissions in Docker containers can be a nightmare so will also need to ensure `/opt/conda` is user writable (apart from  `workspace/numpy`)
- After experimenting with gitpod and SciPy in multiple ways I have found that the best approach is:
1. Create a base image that sets users and permissions and creates the base development environment
2. Create a second image that will be used by Gitpod where we add additional convenience packages and if needed do any prebuilds to avoid relying entirely on gitpod's prebuilds due to lag and to keep the images as slim as possible

**Comment by thomasjpfan** [comment]
Thank you for the comments @trallard !

> It is worth using conda like Ralf mentioned

I agree. The PR was updated to use a conda environment.

> Permissions in Docker containers can be a nightmare so will also need to ensure /opt/conda is user writable

This is why I elected to have conda installed in `/home/gitpod` with user permissions. This avoids all the permissions errors and keeps the Dockerfile simpler.

> Create a base image that sets users and permissions and creates the base development environment

This PR uses `gitpod/workspace-base` as the base image instead of creating our own base image. Since `gitpod/workspace-base` is managed by `gitpod`, any future changes they make will be inherited by numpy's gitpod Dockerfile.

In summary, I aimed to have `Dockerfile`  be as simple as possible, which usually helps with maintenance.

**Comment by thomasjpfan** [comment]
With https://github.com/numpy/numpy/pull/18733/commits/b8a46259a81af5aadb936add7878b5979295c5ee, I removed the github action for building to make this PR easier to review.

**Comment by trallard** [comment]
> This PR uses gitpod/workspace-base as the base image instead of creating our own base image. Since gitpod/workspace-base is managed by gitpod, any future changes they make will be inherited by numpy's gitpod Dockerfile.

This is a good idea - they have been changing a lot of stuff and this `base` image is new (a couple of weeks back they only had a `full` image which was basically everything-but-the-kitchen-sink 👍🏽

File: tools/gitpod/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,59 @@
+FROM gitpod/workspace-base as stage1
+
+COPY --chown=gitpod . /tmp/numpy_repo
+
+# versioneer.py needs a git repo to construct the version
+# We use a multistage build to create a shallow copy of the repo
+RUN git clone --depth 1 file:////tmp/numpy_repo /tmp/numpy
```
Comment (trallard) [inline tools/gitpod/Dockerfile:7]: Do you need to clone the repo? It seems that in the runtime stage you are only using the `environment.yaml` file so might be best to remove multi-stage build and copy the yaml file directly with `COPY` especially if building through GitHub actions

Comment (thomasjpfan) [reply tools/gitpod/Dockerfile:13]: While building the runtime image, numpy is built to populate the cache used by ccache. In principle we should be able to build without a `git` history, but `setup.py` depends on `versioneer.py`, which requires a git history. 

The multistage stage is used to have a shallow clone, which reduces the image size of the final image. 

I updated the PR with detailed comments to help clarify this point.

**Comment by thomasjpfan** [comment]
It takes around ~ 3 minutes from clicking the gitpod button to having a workspace running. Most of the time is for fetching the image. I think this due to the overall [image size](https://hub.docker.com/layers/144479197/thomasjpfan/numpy-gitpod/latest/images/sha256-951f7bc94a990228effa9f7e907aebf136bab1c508b7b54f3a61b1e60e0e5732?context=explore). The layer that installs all the dependencies takes up a little more than half the size of the image.

Are we okay with a 3 minute start time?

**Comment by trallard** [comment]
I think 3 mins is more than reasonable - I am not familiar with the docs build In Numpy but might be worth also doing a build in init so that folks are ready to go straight away

**Comment by rgommers** [comment]
> I think 3 mins is more than reasonable - I am not familiar with the docs build In Numpy but might be worth also doing a build in init so that folks are ready to go straight away

Yes the doc build is quite slow, it's worth adding in the Docker image.

> Most of the time is for fetching the image. I think this due to the overall image size. 

The problem is MKL gets pulled in somehow, even though `environment.yml` specifies `openblas`. Adding an explicit `nomkl` will fix this. That should bring down the image size quite a bit.

**Comment by thomasjpfan** [comment]
> The problem is MKL gets pulled in somehow, even though environment.yml specifies openblas. Adding an explicit nomkl will fix this. That should bring down the image size quite a bit.

Adding `nomkl` to the `environment.yml` worked. Now it takes around 95 seconds to get to a workspace!

> Yes the doc build is quite slow, it's worth adding in the Docker image.

The numpy repo used by the Docker image is different from the numpy `gitpod` clones into `/workspace/numpy`. We can build the docs in the Docker image, but `gitpad`'s clone would not be able to use the sphinx cache when we build docs again. I think ccache works because ccache checks file contents and sphinx checks the mtime. Using ccache is nice for forks because they do not need prebuilds active. The Docker image already has ccache warmed up for building numpy.

I wouldn't want to auto build the docs on startup for now, because a contributor may be fixing a bug that does not require the docs to be build and docs building takes a long time.

The workflow I see for contributors is:

1. Create a fork of numpy
2. Open gitpod with their fork as the url: `https://gitpod.io/#https://github.com/$USERNAME/numpy/tree/main`. The `origin` and `upstream` remotes will be set up correctly here.
3. Numpy will build on startup, which is quick because of ccache.
4. Create a new branch to develop on.
5. Push to their fork.
6. Open PR to numpy.

**Comment by rgommers** [comment]
That makes sense, seems like a good workflow.

> I wouldn't want to auto build the docs on startup for now, because a contributor may be fixing a bug that does not require the docs to be build and docs building takes a long time.

Agreed, that's not a good idea. What SciPy does is build the docs in CI and include the doc/build/ folder into the image. The approach here is a little different; seems like a good idea to try both and see which is nicer.

**Comment by thomasjpfan** [comment]
> Agreed, that's not a good idea. What SciPy does is build the docs in CI and include the doc/build/ folder into the image.

The SciPy approach is effective as long as contributors turn on prebuilds on their fork and wait for gitpod to build the Dockerfile. This wait includes building scipy + building docs. In the end, you get a fully built SciPy and the docs.

The approach in this PR gets to a built NumPy in 1.6 minutes but no docs. A compromise would be to build the docs when the workspace first launches. This way a contributor gets to browse the codebase while the docs build in the background.

A follow up to this PR would be to add a github action to build the Dockerfile, upload it to docker hub, and have `.gitpod.yml` point to `numpy/numpy-gitpod:latest`.

**Comment by rgommers** [comment]
> The SciPy approach is effective as long as contributors turn on prebuilds on their fork and wait for gitpod to build the Dockerfile. This wait includes building scipy + building docs. In the end, you get a fully built SciPy and the docs.

That's not the case. Turning on prebuild on forks is not a good idea in general (just like turning on CI on forks is a waste typically). The way it works for SciPy is:
- You go to https://www.gitpod.io/#https://github.com/scipy/scipy (the docs should still gain a prominent link or button pointing to it)
- This pulls in a Docker image with an in-place build _and_ a doc build already done
- If you don't have commit rights, then Gitpod switches to your fork of SciPy (creating it if it doesn't exist yet) and names your fork `origin` and the main repo `upstream
- Contributors can start developing as soon as the workspace appears

**Review by rgommers — APPROVED** [review]
Let's give this a try, we can iterate as needed. Thanks Thomas!

**Comment by thomasjpfan** [comment]
> This pulls in a Docker image with an in-place build and a doc build already done

I just tried gitpod.io/#https://github.com/scipy/scipy out. It mostly works as as you said: SciPy is built in place and docs are there.

> If you don't have commit rights, then Gitpod switches to your fork of SciPy (creating it if it doesn't exist yet) and names your fork origin and the main repo `upstream

For me (without commit rights), I only have `origin` setup pointing to `scipy/scipy` and would need to adjust the remotes myself. Maybe Gitpod changed something recently?

XREF: https://community.gitpod.io/t/how-to-specify-users-fork-instead-of-main-repository/639/6

**Comment by thomasjpfan** [comment]
Thank you working with me through this PR! I am always looking for workflows that makes it easier to contribute.

**Comment by rgommers** [comment]
> For me (without commit rights), I only have `origin` setup pointing to `scipy/scipy` and would need to adjust the remotes myself. Maybe Gitpod changed something recently?
>
> XREF: https://community.gitpod.io/t/how-to-specify-users-fork-instead-of-main-repository/639/6

Seems like they did - wish I had a better memory ....

> Thank you working with me through this PR! I am always looking for workflows that makes it easier for contributors.

Me too:) Pretty excited about being able to use Gitpod. 

In early May we have a WiMLDS NumPy sprint coming up where we intend to use Gitpod. Before then I'd like to experiment a bit with https://visualstudio.microsoft.com/services/live-share/ within Gitpod - we really need something like that to be able to help new contributors in a sprint when they get stuck.

**Comment by rgommers** [comment]
> In early May we have a WiMLDS NumPy sprint coming up where we intend to use Gitpod. Before then I'd like to experiment a bit with https://visualstudio.microsoft.com/services/live-share/ within Gitpod - we really need something like that to be able to help new contributors in a sprint when they get stuck.

https://www.gitpod.io/docs/sharing-and-collaboration is not updated for the Theia to VS Code move, but it says you can share running workspaces - that doesn't include the editor as in live share, but it will enable looking at the state of the repo and help people get unstuck.
